codex/validation.nim

```diff
@@ -0,0 +1,57 @@
+import std/sets
+import pkg/chronos
+import pkg/chronicles
+import ./market
+import ./clock
+
+export market
+export sets
+
+type
+  Validation* = ref object
+    slots*: HashSet[SlotId]
+    clock: Clock
+    market: Market
+    subscriptions: seq[Subscription]
+    running: Future[void]
+
+proc new*(_: type Validation, clock: Clock, market: Market): Validation =
+  Validation(clock: clock, market: market)
+
+proc subscribeSlotFilled(validation: Validation) {.async.} =
+  proc onSlotFilled(requestId: RequestId, slotIndex: UInt256) =
+    validation.slots.incl(slotId(requestId, slotIndex))
+  let subscription = await validation.market.subscribeSlotFilled(onSlotFilled)
+  validation.subscriptions.add(subscription)
+
+proc subscribeSlotFreed(validation: Validation) {.async.} =
+  proc onSlotFreed(slotId: SlotId) =
+    validation.slots.excl(slotId)
+  let subscription = await validation.market.subscribeSlotFreed(onSlotFreed)
+  validation.subscriptions.add(subscription)
+
+proc run(validation: Validation) {.async.} =
+  try:
+    while true:
+      var ended: HashSet[SlotId]
+      for slotId in validation.slots:
+        let state = await validation.market.slotState(slotId)
+        if state != SlotState.Filled:
+          ended.incl(slotId)
+      validation.slots.excl(ended)
+      await sleepAsync(1.seconds) # TODO: wait for next period
+  except CancelledError:
+    discard
+  except CatchableError as e:
+    error "Validation failed", msg = e.msg
```

Comment:
Are there specific error types that we could catch that may give better error info in the log and provide a better debugging story?

As an example,
```nim
try:
  await validatorContracts.start()
except JsonRpcProviderError as e:
  error "RPC provider error, validation failed", msg = e.msg
except CatchableError as e: # catchall
  error "Validation failed", msg = e.msg
```

Reply:
I don't think that that's a good idea; JsonRpcProviderError is something very deep inside the nim-ethers library, and I don't want the validator to know about it. That's why we have the Market abstraction.

Reply:
Ok, `JsonRpcProviderError` is not the right type then, but we may be able to scratch out some better error info with an error type indicating there was an error coming back from the RPC interface. Perhaps `ProviderError` or `EthersError`?

Reply:
`ProviderError` and `EthersError` are also nim-ethers specific errors, which would tie the validator to a specific implementation of Market. If we really want to have a more specific log message, and we don't want to break the Market abstraction, then we'd have to wrap every method in Market into a `try:except` that catches the underlying error and wraps it in a `MarketError`. But I'm not sure how useful that would be.

Reply:
> But I'm not sure how useful that would be.

It would allow us to recover from certain errors that are maybe not necessarily errors. For example, assuming we were to wrap just the market-specific call in a try/except, we could catch reverts without completely failing the validation run. I'm seeing a lot of "Validation timed out" reverts, which wouldn't necessarily mean we want to stop validating, but instead it would be a good error to log for debug purposes.

Reply:
Ignore this for now, it's more distracting than it needs to be for the bigger picture :)